techlibs/xilinx/xilinx_dffopt.cc

```diff
@@ -0,0 +1,350 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2012  Clifford Wolf <clifford@clifford.at>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/sigtools.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+typedef std::pair<Const, std::vector<SigBit>> LutData;
+
+// Compute a LUT implementing (select ^ select_inv) ? alt_data : data.  Returns true if successful.
+bool merge_lut(LutData &result, const LutData &data, const LutData select, bool select_inv, SigBit alt_data, int max_lut_size) {
+	// First, gather input signals.
+	result.second = data.second;
+	int idx_alt = -1;
+	if (alt_data.wire) {
+		// Check if we already have it.
+		for (int i = 0; i < GetSize(result.second); i++)
+			if (result.second[i] == alt_data)
+				idx_alt = i;
+		// If not, add it.
+		if (idx_alt == -1) {
+			idx_alt = GetSize(result.second);
+			result.second.push_back(alt_data);
+		}
+	}
+	std::vector<int> idx_sel;
+	for (auto bit : select.second) {
+		int idx = -1;
+		for (int i = 0; i < GetSize(result.second); i++)
+			if (result.second[i] == bit)
+				idx = i;
+		if (idx == -1) {
+			idx = GetSize(result.second);
+			result.second.push_back(bit);
```

Comment:
Another late thought I had: this function adds control signals to the last input of a LUT, right? The last input of a LUT is also the fastest one, so would you be slowing down the timing of all the existing `D` path in favour of these control? If so, would it make more sense to put it at the beginning since these control inputs are pretty slow to begin with?

Reply:
I don't think it matters at this point — the P&R tool permutes LUT inputs anyhow, so if our arrangement is an obstacle to meeting timing, it'll be automatically fixed.

Reply:
Not sure I agree with your statement entirely. My view is that post-techmapping we should take care in minimising timing impact of any physical optimisations that are made. For one, on the Yosys roadmap is for it to gain static timing analysis capabilities, and it would be unnecessary for this analysis to have to do LUT input permutations. Secondly, LUT input permutations may be done as late as at the routing stage, in which case any decisions made prior to this could be made based on less-than-optimal arrangement.

What do you think?

Reply:
I don't consider any of these effects to be very important (any mistakes we make about timing will probably be minor compared to the interconnect effect), but I'll do it — the change doesn't seem too hard.